File: npmjs/pray-cli/src/cli/commands/init-prayer-manifest.ts
import { PrayError } from "../../errors.js";
import {
  DEFAULT_PATH_SOURCE_DIRECTORY,
  DEFAULT_PATH_SOURCE_NAME,
} from "../../local-prayer.js";
import { defaultManifestPath } from "../../lockfile/paths.js";
import { parseManifest, readManifestText } from "../../manifest/index.js";
import type { Manifest, ManifestSource } from "../../manifest/types.js";
import { writeProjectFile } from "../../transaction/index.js";

export function parsePrayerInitArguments(argumentsList: string[]): {
  name?: string;
  directory?: string;
} {
  let name: string | undefined;
  let directory: string | undefined;
  for (let index = 0; index < argumentsList.length; index += 1) {
    const argument = argumentsList[index];
    if (argument === undefined) {
      continue;
    }
    if (argument === "--path") {
      const value = argumentsList[index + 1];
      if (!value || value.startsWith("-")) {
        throw PrayError.usage("--path requires a directory");
      }
      directory = value;
      index += 1;
    } else if (argument.startsWith("-")) {
      throw PrayError.unsupported(`unexpected prayer argument: ${argument}`);
    } else if (!name) {
      name = argument;
    } else {
      throw PrayError.unsupported(`unexpected prayer argument: ${argument}`);
    }
  }
  return { name, directory };
}

export function resolveLocalPrayerDirectory(requested?: string): {
  name: string;
  directory: string;
} {
  const manifest = parseManifest(readManifestText(defaultManifestPath()));
  const pathSources = pathSourcesFor(manifest);
  if (requested !== undefined) {
    const directory = requested.trim();
    if (directory.length === 0) {
      throw PrayError.usage("--path requires a directory");
    }
    const matching = pathSources.find((source) => source.url === directory);
    if (matching) {
      return { name: matching.name, directory: matching.url };
    }
    const unique = uniquePathSource(pathSources);
    if (unique) {
      throw PrayError.usage(`path source already uses ${unique.url}`);
    }
    const name = ensurePathSource(directory);
    return { name, directory };
  }
  if (pathSources.length === 0) {
    const name = ensurePathSource(DEFAULT_PATH_SOURCE_DIRECTORY);
    return { name, directory: DEFAULT_PATH_SOURCE_DIRECTORY };
  }
  const unique = uniquePathSource(pathSources);
  if (unique) {
    return { name: unique.name, directory: unique.url };
  }
  throw PrayError.usage("say which directory with --path");
}

export function declareLocalPrayer(packageName: string): void {
  const manifestPath = defaultManifestPath();
  const manifestText = readManifestText(manifestPath);
  const manifest = parseManifest(manifestText);
  if (
    manifest.packages.some((packageEntry) => packageEntry.name === packageName)
  ) {
    return;
  }
  const statement = `pray "${packageName}"`;
  const updated =
    insertIntoFirstCompose(manifestText, statement) ??
    insertTopLevelStatement(manifestText, statement);
  writeProjectFile(manifestPath, updated);
}

function pathSourcesFor(manifest: Manifest): ManifestSource[] {
  return manifest.sources.filter((source) => source.kind === "path");
}

function uniquePathSource(
  pathSources: ManifestSource[],
): ManifestSource | undefined {
  return pathSources.length === 1 ? pathSources[0] : undefined;
}

function ensurePathSource(directory: string): string {
  const manifestPath = defaultManifestPath();
  const manifestText = readManifestText(manifestPath);
  const manifest = parseManifest(manifestText);
  const existing = manifest.sources.find(
    (source) => source.kind === "path" && source.url === directory,
  );
  if (existing) {
    return existing.name;
  }
  const name = unusedSourceName(manifest, directory);
  const statement = `source "${name}", path: "${directory}"`;
  writeProjectFile(
    manifestPath,
    insertSourceStatement(manifestText, statement),
  );
  return name;
}

function unusedSourceName(manifest: Manifest, directory: string): string {
  const taken = new Set(manifest.sources.map((source) => source.name));
  if (!taken.has(DEFAULT_PATH_SOURCE_NAME)) {
    return DEFAULT_PATH_SOURCE_NAME;
  }
  const segments = directory
    .split(/[/\\]/)
    .filter((segment) => segment.length > 0);
  const fallback = segments.at(-1) ?? DEFAULT_PATH_SOURCE_NAME;
  if (taken.has(fallback)) {
    throw PrayError.manifest(`source name ${fallback} is already used`);
  }
  return fallback;
}

function insertSourceStatement(text: string, statement: string): string {
  const lines = text.split(/\r?\n/);
  let insertionIndex = 0;
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    const line = lines[index];
    if (line?.trimStart().startsWith("source ")) {
      insertionIndex = index + 1;
      break;
    }
    if (index === 0) {
      const prayfile = lines.findIndex((line) =>
        line.trimStart().startsWith("prayfile "),
      );
      insertionIndex = prayfile >= 0 ? prayfile + 1 : 0;
    }
  }
  lines.splice(insertionIndex, 0, statement);
  return joinManifestLines(lines);
}

function insertTopLevelStatement(text: string, statement: string): string {
  const lines = text.split(/\r?\n/);
  const insertionIndex = lines.findIndex((line) => {
    const trimmed = line.trimStart();
    return trimmed.startsWith("local ") || trimmed.startsWith("render ");
  });
  lines.splice(
    insertionIndex >= 0 ? insertionIndex : lines.length,
    0,
    statement,
  );
  return joinManifestLines(lines);
}

function insertIntoFirstCompose(
  text: string,
  statement: string,
): string | undefined {
  const lines = text.split(/\r?\n/);
  const index = lines.findIndex((line) => {
    const trimmed = line.trimStart();
    return trimmed.startsWith("compose ") && trimmed.endsWith(" do");
  });
  if (index < 0) {
    return undefined;
  }
  const composeLine = lines[index];
  if (composeLine === undefined) {
    return undefined;
  }
  const indent = composeLine.match(/^\s*/)?.[0] ?? "";
  lines.splice(index + 1, 0, `${indent}  ${statement}`);
  return joinManifestLines(lines);
}

function joinManifestLines(lines: string[]): string {
  return `${lines.join("\n").replace(/\n*$/, "")}\n`;
}

File: rubygems/pray-cli/lib/pray/cli/commands/init_prayer_manifest.rb
# frozen_string_literal: true

module Pray
  module CLI
    def resolve_local_prayer_directory(requested)
      manifest = Pray.parse_manifest(Pray.read_manifest_text(manifest_path))
      path_sources = path_sources_for(manifest)
      if requested
        directory = requested.to_s.strip
        raise Error.usage("--path requires a directory") if directory.empty?

        matching = path_sources.find { |source| source.url == directory }
        return [matching.name, matching.url] if matching
        if path_sources.length == 1
          raise Error.usage("path source already uses #{path_sources.first.url}")
        end

        name = ensure_path_source(directory)
        return [name, directory]
      end

      case path_sources.length
      when 0
        name = ensure_path_source(DEFAULT_PATH_SOURCE_DIRECTORY)
        [name, DEFAULT_PATH_SOURCE_DIRECTORY]
      when 1
        [path_sources.first.name, path_sources.first.url]
      else
        raise Error.usage("say which directory with --path")
      end
    end

    def declare_local_prayer(package_name)
      manifest_text = Pray.read_manifest_text(manifest_path)
      manifest = Pray.parse_manifest(manifest_text)
      return if manifest.packages.any? { |package| package.name == package_name }

      statement = "pray \"#{package_name}\""
      updated = insert_into_first_compose(manifest_text, statement) ||
        insert_manifest_statement(manifest_text, statement)
      Transaction.write_file(manifest_path, updated)
    end

    def path_sources_for(manifest)
      manifest.sources.select { |source| source.kind == "path" }
    end

    def ensure_path_source(directory)
      manifest_text = Pray.read_manifest_text(manifest_path)
      manifest = Pray.parse_manifest(manifest_text)
      existing = manifest.sources.find { |source| source.kind == "path" && source.url == directory }
      return existing.name if existing

      name = unused_source_name(manifest, directory)
      statement = "source \"#{name}\", path: \"#{directory}\""
      Transaction.write_file(manifest_path, insert_source_statement(manifest_text, statement))
      name
    end

    def unused_source_name(manifest, directory)
      taken = manifest.sources.map(&:name)
      return DEFAULT_PATH_SOURCE_NAME unless taken.include?(DEFAULT_PATH_SOURCE_NAME)

      fallback = directory.to_s.split(%r{[/\\]}).reject(&:empty?).last || DEFAULT_PATH_SOURCE_NAME
      raise Error.manifest("source name #{fallback} is already used") if taken.include?(fallback)

      fallback
    end

    def insert_source_statement(text, statement)
      lines = text.lines.map(&:chomp)
      last_source = lines.rindex { |line| line.lstrip.start_with?("source ") }
      insertion_index = if last_source
        last_source + 1
      else
        prayfile = lines.index { |line| line.lstrip.start_with?("prayfile ") }
        prayfile ? prayfile + 1 : 0
      end
      lines.insert(insertion_index, statement)
      join_manifest_lines(lines)
    end

    def insert_into_first_compose(text, statement)
      lines = text.lines.map(&:chomp)
      index = lines.index do |line|
        trimmed = line.lstrip
        trimmed.start_with?("compose ") && trimmed.end_with?(" do")
      end
      return unless index

      indent = lines[index][/\A\s*/].to_s
      lines.insert(index + 1, "#{indent}  #{statement}")
      join_manifest_lines(lines)
    end

    def join_manifest_lines(lines)
      output = lines.join("\n")
      output += "\n" unless output.end_with?("\n")
      output
    end
  end
end
